drivers/dac/dac_max22017.c

```diff
@@ -448,11 +459,298 @@ static void max22017_int_worker(struct k_work *work)
 	ret = FIELD_GET(MAX22017_GEN_INT_GPI_INT, gen_int);
 	if (ret) {
 		LOG_INF("GPI Interrupt: %d", ret);
+#ifdef CONFIG_GPIO_MAX22017
+		uint16_t gpi_sta, lsb;
+
+		ret = max22017_reg_read(dev, MAX22017_GEN_GPI_INT_STA_OFF, &gpi_sta);
+		if (ret) {
+			goto fail;
+		}
+
+		/* Aggregate both positive and negative edge together */
+		gpi_sta = FIELD_GET(MAX22017_GEN_GPI_INT_GPI_NEG_EDGE_INT, gpi_sta) |
+			  FIELD_GET(MAX22017_GEN_GPI_INT_GPI_POS_EDGE_INT, gpi_sta);
+		while (gpi_sta) {
+			lsb = LSB_GET(gpi_sta);
+			gpio_fire_callbacks(&data->callbacks_gpi, dev, lsb);
+			gpi_sta &= ~lsb;
+		}
+#endif
 	}
 fail:
 	k_sem_give(&data->lock);
 }
 
+#ifdef CONFIG_GPIO_MAX22017
+
+#ifdef CONFIG_GPIO_MAX22017_INT_QUIRK
+void isr_quirk_handler(struct k_timer *int_quirk_timer)
+{
+	int ret;
+	struct max22017_data *data = k_timer_user_data_get(int_quirk_timer);
+	k_sem_take(&data->lock, K_FOREVER);
+
+	ret = k_work_submit(&data->int_work);
+	if (ret < 0) {
+		LOG_WRN("Could not submit int work: %d", ret);
+	}
+
+	k_sem_give(&data->lock);
+}
+#endif
+
+int adi_max22017_gpio_set_output(const struct device *dev, uint8_t pin, bool initial_value)
```

Comment:
Why do all these GPIO related functions need to live in this file? Can they not be moved to `gpio_max22017.c`?

Reply:
they can live in `gpio_max22017.c` I guess with a bit of re-org.

My rationale was that the `gpio` driver is actually just a "proxy" to the parent dac device where everything happens.
So the spi and all internal data structure of the dac driver are used to perform the gpio related functions.

Reply:
Have you considered implementing an MFD base driver instead, having the DAC and GPIO drivers as children/consumers?

Reply:
@henrikbrixandersen : The current design have been "inspired" by `drivers/adc/adc_ads114s0x.c` 
I've seen the MFD subsystem afterwards and wasn't sure it was worth using it.

I can change the design if you think using an MFD is a better choice.

Reply:
I agree that the MFD approach is better.

Reply:
I'm looking into reworking this into an MFD then.
Dumb question: how many PRs would you like to see? 3 ? one for MFD, one for DAC, one for GPIO ?

Reply:
In this case, just one.